File: uis/client/mobile_app/lib/services/enaira_service.dart
import '../models/enaira_wallet.dart';
import '../config/app_config.dart';
import 'api_service.dart';

class EnairaService {
  final ApiService _apiService;

  EnairaService(this._apiService);

  static const String _base = AppConfig.enairaCbdcEndpoint;

  Future<List<EnairaWallet>> getWallets({int page = 1, int limit = 20, String? status}) async {
    final params = <String, dynamic>{'page': page, 'limit': limit};
    if (status != null) params['status'] = status;

    final response = await _apiService.get('$_base/v1/wallets', queryParameters: params);
    final data = response.data;
    final rawItems = data is Map ? (data['items'] ?? data['data'] ?? []) : data;
    if (rawItems is! List) return [];
    return rawItems
        .whereType<Map<String, dynamic>>()
        .map(EnairaWallet.fromJson)
        .toList();
  }

  Future<EnairaWallet> getWalletById(String id) async {
    final response = await _apiService.get('$_base/v1/wallets/$id');
    return EnairaWallet.fromJson(Map<String, dynamic>.from(response.data));
  }

  Future<EnairaWallet> createWallet(Map<String, dynamic> body) async {
    final response = await _apiService.post('$_base/v1/wallets', data: body);
    return EnairaWallet.fromJson(Map<String, dynamic>.from(response.data));
  }

  Future<Map<String, dynamic>> fundWallet({
    required String walletId,
    required double amount,
  }) async {
    final response = await _apiService.post('$_base/v1/wallets/$walletId/fund', data: {
      'amount': amount,
    });
    return Map<String, dynamic>.from(response.data);
  }

  Future<Map<String, dynamic>> transfer({
    required String fromWalletId,
    required String toWalletId,
    required double amount,
    String? narration,
  }) async {
    final response = await _apiService.post('$_base/v1/transfers', data: {
      'from_wallet_id': fromWalletId,
      'to_wallet_id': toWalletId,
      'amount': amount,
      if (narration != null) 'narration': narration,
    });
    return Map<String, dynamic>.from(response.data);
  }

  Future<Map<String, dynamic>> getStats() async {
    final response = await _apiService.get('$_base/v1/stats');
    return Map<String, dynamic>.from(response.data);
  }

  Future<List<Map<String, dynamic>>> getTransactions(String walletId, {int page = 1, int limit = 20}) async {
    final response = await _apiService.get(
      '$_base/v1/wallets/$walletId/transactions',
      queryParameters: {'page': page, 'limit': limit},
    );
    final data = response.data;
    final rawItems = data is Map ? (data['items'] ?? data['data'] ?? []) : data;
    if (rawItems is! List) return [];
    return rawItems.whereType<Map<String, dynamic>>().toList();
  }
}

File: uis/client/web2/src/services/enaira_service.ts
import { apiService } from './api_service';

const BASE = '/enaira-cbdc/v1';

export interface ENairaWallet {
  id: string;
  wallet_address?: string;
  wallet?: string;
  tier?: string;
  balance?: number | string;
  status?: string;
  created_at?: string;
  last_activity?: string;
  lastActivity?: string;
  [key: string]: unknown;
}

export interface ENairaTransfer {
  id: string;
  sender_wallet?: string;
  recipient_wallet?: string;
  amount: number | string;
  status?: string;
  created_at?: string;
  [key: string]: unknown;
}

export interface ENairaStats {
  total_wallets?: number;
  active_wallets?: number;
  total_volume?: number | string;
  tier1_count?: number;
  tier2_count?: number;
  tier3_count?: number;
  [key: string]: unknown;
}

interface ApiResult<T> {
  success: boolean;
  data: T | null;
  message: string;
}

async function call<T>(fn: () => Promise<{ data: unknown }>): Promise<ApiResult<T>> {
  try {
    const res = await fn();
    const payload = res.data as { data?: T; results?: T } | T;
    const data =
      payload && typeof payload === 'object' && ('data' in payload || 'results' in payload)
        ? ((payload as { data?: T; results?: T }).data ?? (payload as { results?: T }).results ?? (payload as T))
        : (payload as T);
    return { success: true, data, message: 'OK' };
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : 'Request failed';
    return { success: false, data: null, message: msg };
  }
}

class ENairaService {
  getWallets() {
    return call<ENairaWallet[]>(() => apiService.get(`${BASE}/wallets`));
  }

  getWalletById(id: string) {
    return call<ENairaWallet>(() => apiService.get(`${BASE}/wallets/${id}`));
  }

  createWallet(data: Record<string, unknown>) {
    return call<ENairaWallet>(() => apiService.post(`${BASE}/wallets`, data));
  }

  fundWallet(id: string, data: Record<string, unknown>) {
    return call<ENairaWallet>(() => apiService.post(`${BASE}/wallets/${id}/fund`, data));
  }

  transfer(data: Record<string, unknown>) {
    return call<ENairaTransfer>(() => apiService.post(`${BASE}/transfers`, data));
  }

  getStats() {
    return call<ENairaStats>(() => apiService.get(`${BASE}/stats`));
  }

  getTransactions() {
    return call<ENairaTransfer[]>(() => apiService.get(`${BASE}/transactions`));
  }
}

export const enairaService = new ENairaService();
